rfcs/0001-polling-mechanism/README.md

```diff
@@ -0,0 +1,259 @@
+
+1. Implement a secure mechanism to track changes in a branch-based planner with GitHub, minimizing the exposure of
+   Kubernetes clusters to potential security risks.
+2. Develop a portable solution that can be adapted for different Git providers (e.g., GitLab, Bitbucket), reducing the
+   complexity of dealing with varying webhook systems or APIs.
+3. Manage the potential drawbacks of the polling mechanism, such as latency and load on the GitHub server, through
+   careful configuration of the polling frequency.
+4. Efficiently handle GitHub API rate limits to prevent the client from being temporarily blocked from making additional
+   requests.
+5. Securely store and manage GitHub tokens within Kubernetes Secrets to mitigate security risks associated with access
+   tokens.
+
+### Non-Goals
+
+1. Creating a real-time update system. While desirable, real-time updates may compromise security and increase system
+   complexity. The proposed polling mechanism will inherently introduce some latency.
+2. Building custom parsing and handling code for each Git provider's webhook messages. Our aim is to standardize
+   interactions with Git providers, rather than dealing with specificities of each provider's webhook system.
+3. Eliminating all load on the GitHub server. Some load is inevitable due to the nature of polling; our goal is to
+   minimize and manage this load rather than eliminate it completely.
+
+## Portability of Polling Mechanism to Other Git Providers
+
+One of the key advantages of using a polling mechanism is the ease of porting this implementation to other Git providers
+such as GitLab, Bitbucket, etc. With the polling mechanism, we can standardize the way we interact with the Git
+providers, reducing the complexity that comes with dealing with different webhook systems or APIs.
+
+Webhook support varies widely among Git providers. Some providers might not support webhooks at all, or they might have
+different ways of setting up and securing webhooks. Moreover, the structure and content of webhook messages can also
+differ significantly between providers, which means the code to parse and handle these messages would need to be written
+and maintained for each provider.
+
+On the contrary, the polling mechanism works largely the same way for any Git provider. We need to make requests to the
+provider's API to fetch the required data, which is typically available through similar RESTful APIs across different
+providers. Therefore, we can easily adapt our polling code to a new provider by changing the API endpoints and adjusting
+for any differences in the API responses.
+
+In addition, the polling mechanism allows us to control the rate of requests, which can be beneficial for dealing with
+rate limits imposed by different providers. With webhooks, the rate of incoming messages is determined by the activity
+on the Git provider, which could potentially overwhelm our system or cause us to exceed rate limits.
+
+## Proposed solution
```

Comment:
For polling, do we need an extra CRD that describes a repo to watch? Something like this:

```yaml
---
apiVersion: infra.contrib.fluxcd.io/v1alpha2
kind: TerraformBranchBasedPlanner
metadata:
  name: helloworld-tf
  namespace: flux-system
spec:
  interval: 1m
  sourceRef:
    kind: GitRepository
    name: helloworld
    namespace: flux-system
```

Reply:
I would go with the simplest implementation first without introducing a new CRD.

Reply:
Hmm... and what would be the "simplest implementation"?

Reply:
A simple polling, PR query, etc without CRD - that's what I meant.

Reply:
Yes, but how the system will know what repositories it has to query for PRs?

Reply:
Would you please create a spike ticket and move this discussion there? It's a bit out of scope for this RFC.

Reply:
Yes, but I don't think it's fully out of scope as it's probably a "drawback" as we have to either create a CRD or iterate over all Terraform resources on the cluster, parse their repo, aggregate the repositories (so no duplicates) and then we can query PRs. Compared to webhook where we know the repository from received payload.

Reply:
> Compared to webhook where we know the repository from received payload.

I mistook your concern then. 

We have that information in the code provided. See example 1.

Reply:
In the provided code we have a hard-coded repository.
```go
prs, _, _ := client.PullRequests.List(ctx, "weaveworks", "tf-controller", nil)
```

Reply:
ConfigMap sounds good.

Reply:
Scoping down to a set of specified repositories would also improve security.

Thank you for bringing this up @yitsushi :+1: